### PR TITLE
Add tests validating 32 byte chainwork checkpoints can be processed once RSKIP454 is activated

### PR DESCRIPTION
**Pull request by nathanieliov**
## Description

Add tests validating 32 byte chainwork checkpoints can be processed once RSKIP454 is activated

## How Has This Been Tested?

Unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Requires Activation Code (Hard Fork)

### CODE REVIEW

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java [hunk 1]
```diff
@@ -192,17 +199,114 @@ void testInitialChainHeadWithoutBtcCheckpoints() throws Exception {
...
 
+        BridgeSupport bridgeSupport = arrangeBridgeSupport(bridgeTestNetConstants,
+            activationsBeforeForks);
+
+        // Force instantiation of blockstore
+        bridgeSupport.getBtcBlockchainBestChainHeight();
+
+        InputStream checkpointsStream = bridgeSupport.getCheckPoints();
+        CheckpointManager manager = new CheckpointManager(bridgeTestNetConstants.getBtcParams(), checkpointsStream);
+        long time = bridgeSupport.getActiveFederation().getCreationTime().toEpochMilli() - 604800L; // The magic number is a substraction CheckpointManager does when getting the checkpoints.
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:210]:
```suggestion
        long checkpointManagerSubstraction = 604800L;
        long time = bridgeSupport.getActiveFederation().getCreationTime().toEpochMilli() - checkpointManagerSubstraction;
```

wdyt?

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java [hunk 2]
```diff
@@ -192,17 +199,114 @@ void testInitialChainHeadWithoutBtcCheckpoints() throws Exception {
     void testInitialChainHeadWithBtcCheckpoints() throws Exception {
         BridgeConstants bridgeTestNetConstants = BridgeTestNetConstants.getInstance();
 
+        BridgeSupport bridgeSupport = arrangeBridgeSupport(bridgeTestNetConstants,
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:202]: lets add tests for mainnet, wdyt?

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java [hunk 3]
```diff
@@ -192,17 +199,114 @@ void testInitialChainHeadWithoutBtcCheckpoints() throws Exception {
...
+            activationsBeforeForks);
+
+        // Force instantiation of blockstore
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:205]: why doing this?

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java [hunk 4]
```diff
@@ -192,17 +199,114 @@ void testInitialChainHeadWithoutBtcCheckpoints() throws Exception {
...
+    }
+
+    @ParameterizedTest
+    @CsvSource(
+        {
+            "12-byte-chainwork.production.checkpoints",
+            "12-byte-chainwork-mix-format.production.checkpoints"
+        }
+    )
+    void testInitialChainHeadWithBtcCheckpoints_whenCheckpointsWith12BytesChainWork_before_RSKIP454_ok(
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:223]: nit:

```suggestion
    void testInitialChainHeadWithBtcCheckpoints_whenCheckpointsWith12BytesChainWork_beforeRSKIP454_ok(
```

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java [hunk 5]
```diff
@@ -192,17 +199,114 @@ void testInitialChainHeadWithoutBtcCheckpoints() throws Exception {
...
+
+        InputStream checkpointsStream = bridgeSupport.getCheckPoints();
+        CheckpointManager manager = new CheckpointManager(bridgeConstants.getBtcParams(), checkpointsStream);
+        long time = bridgeSupport.getActiveFederation().getCreationTime().toEpochMilli() - 604800L; // The magic number is a substraction CheckpointManager does when getting the checkpoints.
+        StoredBlock checkpoint = manager.getCheckpointBefore(time);
+
+        assertEquals(checkpoint.getHeight(), bridgeSupport.getBtcBlockchainBestChainHeight());
+    }
+
+    private BridgeSupport arrangeBridgeSupport(BridgeConstants bridgeTestNetConstants,
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:295]: not a big fan of this. Can't we set the `bridgeSupport` in the general setup?

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 4 above)
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:223]: can we add the arrange, act, assert sections as we've been doing? I find the tests a bit hard to follow

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 2 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:202]: All new tests are for mainnnet.

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 3 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:205]: Doing exactly what's being said. Check this: https://github.com/rsksmart/rskj/blob/8220d123db03376d90b1357ada2d11f65c93de89/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L2573

Basically, we are testing how checkpoints are read and then store in the `RepositoryBtcBlockStoreWithCache`.

Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:205]: my concern is: why is this needed? maybe an explanatory comment to add a bit more context

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 5 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:295]: I need to build the `bridgeSupport` using different bridgeConstants and different activations depending on the use case. This is why I don't put it there, but I'm reorganizing the tests, and let's see if you find it easier to follow.

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 1 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java]: Refactored.

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 2 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java]: Refactored

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 3 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:208]: Refactored

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 4 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java]: done

Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java]: Done

**Review by apancorb — APPROVED** [review]
LGTM

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java [hunk 6]
```diff
@@ -155,73 +175,147 @@ void setUpOnEachTest() {
...
+        InputStream checkpointsStream = bridgeSupport.getCheckPoints();
+        CheckpointManager manager = new CheckpointManager(bridgeConstants.getBtcParams(),
+            checkpointsStream);
+        long genesisFedCreationTimeAdjusted = bridgeSupport.getActiveFederation().getCreationTime().toEpochMilli() - checkpointTimeAdjustmentForPossibleCLockDrift;
+        StoredBlock checkpoint = manager.getCheckpointBefore(genesisFedCreationTimeAdjusted);
+
+        assertEquals(checkpoint.getHeight(), bridgeSupport.getBtcBlockchainBestChainHeight());
+    }
+
+    private void arrangeBridgeSupport() {
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:296]:
```suggestion
    private void arrangeBridgeSupport(bridgeConstants, activations) {
```

wdyt? to avoid doing the
```
bridgeConstants = specificConstants
activations = specificActivations
```

Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:296]: btw, i'd set the bridgeSupport in the general setup, and create it again when needed in specific tests.
In general we've been trying to avoid using these type of methods

Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:296]: This whole class needs a refactor for the other methods that are not being touched. But this is out of the scope of this task. 

Anyway, regarding using this approach for arrangement in a test, I think you may be referencing the name `arrangeBridgeSupport,` which could be different. I accept that. But this is the same approach we've been using lately. I got inspired by changes in the  `svp` project. E.g., https://github.com/rsksmart/rskj/blob/add-federation-context/rskj-core/src/test/java/co/rsk/peg/BridgeSupportSvpTest.java#L1148C18-L1148C49

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java [hunk 7]
```diff
@@ -129,10 +146,13 @@ public class BridgeSupportIT {
...
     private WhitelistStorageProvider whitelistStorageProvider;
     private LockingCapSupport lockingCapSupport;
+    private Repository track;
+
+    private BridgeSupport bridgeSupport;
 
     @BeforeEach
     void setUpOnEachTest() {
-        bridgeConstants = new BridgeRegTestConstants();
+        bridgeConstants = bridgeRegTestConstants;
```
[the remark this answers is not in the document]
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:155]: Updated

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 6 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java]: Added your suggestion

**Review by jeremy-then — APPROVED** [review]
LGTM.

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java [hunk 8]
```diff
@@ -155,88 +174,151 @@ void setUpOnEachTest() {
             LockingCapMainNetConstants.getInstance(),
             signatureCache
         );
+
+        track = createRepository().startTracking();
+        arrangeBridgeSupport(bridgeRegTestConstants, activations);
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:179]: i insist, don't do this. we don't want a method for arranging the bridge support, that's why we created the bridge support builder.

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java [hunk 9]
```diff
@@ -97,6 +104,15 @@
...
 @MockitoSettings(strictness = Strictness.LENIENT)
 public class BridgeSupportIT {
+
+    private static final long dayInSeconds = 24 * 60 * 60;
+    private static final long weekInSeconds = 7 * dayInSeconds;
+    /**
+     * Time to use in CheckpointManager adjust checkpoint backwards by a week to account for possible clock drift in the block headers.
+     * For more detail please see {@link CheckpointManager#checkpoint(NetworkParameters, InputStream, co.rsk.bitcoinj.store.BtcBlockStore, long)}
+     */
+    private static final long checkpointTimeAdjustmentForPossibleCLockDrift = weekInSeconds;
```
[the remark this answers is not in the document]
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:114]: btw, this variable is just being used [here](https://github.com/rsksmart/rskj/pull/2865/files#diff-e62e3dd078a0552450842b3083d14b35b3514536b8d4e74060bc520696eca5e8R278).
I'd remove it and use weekInSeconds directly, and move the explanatory comment to where its being used.

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java [hunk 10]
```diff
@@ -247,30 +329,16 @@ void feePerKbFromStorageProvider() {
...
-        BridgeSupport bridgeSupport = bridgeSupportBuilder
-            .withBridgeConstants(bridgeConstants)
-            .withProvider(provider)
-            .withRepository(track)
-            .withBtcBlockStoreFactory(btcBlockStoreFactory)
-            .withFederationSupport(federationSupport)
-            .build();
+        // arrange
+        activations = activationsBeforeForks;
+        bridgeRegTestConstants = bridgeRegTestConstants;
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:334]: ?

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java [hunk 11]
```diff
@@ -247,30 +329,16 @@ void feePerKbFromStorageProvider() {
...
+        arrangeBridgeSupport(bridgeRegTestConstants, activations);
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:335]:
```suggestion
        arrangeBridgeSupport(bridgeRegTestConstants, activationsBeforeForks);
```

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 8 above)
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:179]: btw, can we use mainnet for general setup?

Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java:179]: What's your proposal for avoid recreating the bridge support when we need an specific BridgeSupport on certain test cases?

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 10 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java]: Updated

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 8 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java]: Done. Got rid `arrangeBridgeSupport`.

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java]: done

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java (hunk 11 above)
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportIT.java]: Done

**Comment by nathanieliov** [comment]
Closing this PR in favor of reviewing these changes in #2872
